src/Storages/System/StorageSystemZooKeeper.cpp

```diff
@@ -424,9 +424,35 @@ void ReadFromSystemZooKeeper::applyFilters()
     paths = extractPath(getFilterNodes().nodes, context, context->getSettingsRef().allow_unrestricted_reads_from_keeper);
 }
 
+/// Executes a request to Keeper and retries it in case of expired sessions and disconnects
+template <typename Result, typename Operation>
+static Result runWithReconnects(Operation && operation, ContextPtr context, QueryStatusPtr query_status)
+{
+    constexpr int max_retries = 20; /// Limit retries by some reasonable number to avoid infinite loops
```

Comment:
Why isn't this using `ZooKeeperRetriesControl`? It's better to adapt that if we need to rather than re-implement retries

Reply:
Ok
Do you think we should move ZooKeeperRetriesControl from Storages/MergeTree to Common?

Reply:
Absolutely. Ideally we should be using `ZooKeeperRetriesControl`  + `ZooKeeperWithFaultInjectionPtr` everywhere  so we handle both injections and retries.